initiate eigenlayer project, models and sources

bowenli86 opened the pull request:
### Description:

- initiate eigenlayer project and eigenlayer source
- add several models, covering
  - tvl
  - inflow/outflow
  - tvd
  - avs
  - operator
  - rewards v1
  - programmatic incentive
  - strategies


---
quick links for more information:
- [README.md](https://github.com/duneanalytics/spellbook/blob/main/README.md)
- [spellbook docs](https://github.com/duneanalytics/spellbook/tree/main/docs)
- [CONTRIBUTING.md](https://github.com/duneanalytics/spellbook/blob/main/CONTRIBUTING.md)

bowenli86 commented:
I have read the CLA Document and I hereby sign the CLA

bowenli86 commented:
@jeff-dude @0xRobin

Hosuke commented on dbt_subprojects/daily_spellbook/models/eigenlayer/v2_withdrawal_queued_flattened.sql:
```diff
@@ -0,0 +1,51 @@
+            evt_block_time,
+            evt_block_number,
+            withdrawalRoot,
+            JSON_PARSE(withdrawal) AS parsed_withdrawal
+        FROM
+            {{ source('eigenlayer_ethereum', 'DelegationManager_evt_WithdrawalQueued') }}
```
May be incremental?

bowenli86 replied:
i tested it, it only take like 2s in dune trino, so it prob not worth it imho

0xRobin reviewed (commented):
Hey @bowenli86, thanks for starting to build on spellbook! Very exciting to have some official EigenLayer eyes on this.  
We will need some more context on your vision so we can help you further.

Ideally we build the end spells as simple generic tables that are able to serve a multitude of wizards needs. 

- What end goal tables do you have in mind to expose as a spell?
- What's your usecase for needing these tables, what are some other example usecases that wizards will be able to use the spells for?
- What data do they currently have available (and thus what gap are we trying to fill with the spell)?

bowenli86 commented:
@0xRobin thank you

> What end goal tables do you have in mind to expose as a spell?
> What's your usecase for needing these tables, what are some other example usecases that wizards will be able to use the spells for?

none of Eigenlayer TVL boards on Dune right now are correct, b/c the calculation needs complex processing by joining multiple events, and decipher them correctly, etc

for the first phase, we will expose a few key metrics like TVL, operator<>AVS relationships, and rewards. I will decode events (bronze) and add static data to Dune, build several intermediary (silver) tables, and final gold tables for TVL, operator<>AVS relationships, and rewards


> What data do they currently have available (and thus what gap are we trying to fill with the spell)?

gaps as mentioned above. all tables i add to dune will fill the gap

eg. flattening withdrawalqueued events as a table is first step to decipher events to correct bronze tables which will contribute to futher development on top

bowenli86 commented:
@0xRobin @0xBoxer @jeff-dude can you help review?

jeff-dude commented:
> @0xRobin @0xBoxer @jeff-dude can you help review?

thank you for the patience while the team had some time off. @0xRobin will help continue review this upcoming week 🙏

0xRobin commented on dbt_subprojects/daily_spellbook/models/eigenlayer/v2_withdrawal_queued_flattened.sql:
```diff
@@ -0,0 +1,51 @@
+    config(
+        schema = 'eigenlayer',
+        alias = 'v2_withdrawal_queued_flattened',
+        post_hook='{{ expose_spells(\'["ethereum"]\',
+                                    "project",
+                                    "eigenlayer",
+                                    \'["bowenli"]\') }}'
+    )
+}}
```
we should remove this given it's a staging view model, which will be used as a building block for later models. We'd only want the final models to be exposed as spells.

bowenli86 replied:
this will be for both internal building block and external facing model, eg. to directly see who has queued withdraws

bowenli86 replied:
all of our models will be both internal and external models

0xRobin reviewed (changes requested):
@bowenli86 You can keep building on this PR until the final models are ready.
If you need the intermediate table to test data and quality, please use the temporary tables provided by the CI.

bowenli86 commented:
> @bowenli86 You can keep building on this PR until the final models are ready. If you need the intermediate table to test data and quality, please use the temporary tables provided by the CI.

@0xRobin there will prob be 10-20 models in V1, do you mean you guys prefer everything in a giant PR V.S. breaking them down to individual PRs and building things out incrementally?

i felt the latter approach is more scalable and easy to review. i can do former one if that's the convention in this repo

0xRobin commented:
For new lineages:
as much as possible in the initial PR, this gives us maximum context and allows us to provide the most optimal feedback on how to structure everything. If there are any doubts during this process on how anything should be set up, you're definitely free to raise questions or request intermediate reviews. It's important for us to understand the final models to be able to provide proper guidance on how the intermediate tables should be best set up.

For modifications or extensions of existing lineages:
Smaller PRs for easy and fast reviews.

Does that make sense? This also allows you to keep working on this without being blocked by our reviews for each new intermediate model.

bowenli86 commented:
i think this is a pretty good state for initialization with 10+ models, @0xRobin can you help review?

0xRobin reviewed (changes requested):
can we remove the post_hook for any models that aren't "gold" (final models in the lineage)?

We want to minimize the models listed on dune in the data explorer to help analyst find the best models to work with from a small subset. 
The models that aren't exposed can still be queried on Dune, they just won't be visible in the data explorer.

bowenli86 commented:
> can we remove the post_hook for any models that aren't "gold" (final models in the lineage)?
> 
> We want to minimize the models listed on dune in the data explorer to help analyst find the best models to work with from a small subset. The models that aren't exposed can still be queried on Dune, they just won't be visible in the data explorer.

done, can you take another look?

bowenli86 commented:
@0xRobin can you help review?

0xRobin commented on dbt_subprojects/daily_spellbook/models/eigenlayer/eigenlayer_strategy_category.sql:
```diff
@@ -0,0 +1,32 @@
+  strategy
+from (
+values
+  ('eth lst','stETH','0x93c4b944d05dfe6df7645a86cd2206016c51564d'),
```
you should make the addresses varbinary instead of strings (remove the quotes)

bowenli86 commented:
@0xRobin thanks, addressed feedback, can you review?

btw, can you review all changes and provide feedback at once so i can address them at once? the PR has been open for 4 weeks, it's been frustrating it's moving slow. We'd definitely want this in ASAP so we can start using it on Dune

bowenli86 commented:
@0xRobin thank you for the approval!

does the workflow also need to be approved? i can't seem to merge without it passing and it's been pending

<img width="1492" alt="Screenshot 2025-02-24 at 11 46 56 AM" src="https://github.com/user-attachments/assets/fdc610f8-c0ee-4125-9348-11bab227f5a6" />
<img width="1012" alt="Screenshot 2025-02-24 at 11 45 22 AM" src="https://github.com/user-attachments/assets/651de1fe-2ed9-4ad5-8989-a0ccabe52e1e" />

jeff-dude commented:
only us as maintainers can approve and run workflows. this typically only happens on first time contributor PRs. moving forward, you shouldn't hit that block. as for merging, that will also only fall on maintainers.

i will give a final review and prep for merge shortly.

jeff-dude commented:
thank you for the patience. i have pushed a few commits, to try to help expedite. these are mostly best practice items / reusing new utils tables for time series.
- use `_projects` directory, to help keep `eigenlayer` models near the top and easier to find. most below that are not in this directory are legacy and we just haven't taken time to clean / organize completely
- use blockchain in schema names, then corresponding changes to files to match. these are all ethereum, so using that chain name. this also allows flexibility in the future to add other chains without naming conflict
- as mentioned, use new `utils.days` rather than rebuild yourself

bowenli86 commented:
> thank you for the patience. i have pushed a few commits, to try to help expedite. these are mostly best practice items / reusing new utils tables for time series.
> 
> * use `_projects` directory, to help keep `eigenlayer` models near the top and easier to find. most below that are not in this directory are legacy and we just haven't taken time to clean / organize completely
> * use blockchain in schema names, then corresponding changes to files to match. these are all ethereum, so using that chain name. this also allows flexibility in the future to add other chains without naming conflict
> * as mentioned, use new `utils.days` rather than rebuild yourself

make sense, thanks!

jeff-dude commented:
okay, this all looks good to me. can you do me a favor, and query all 15 views built in CI process [here](https://github.com/duneanalytics/spellbook/actions/runs/13509267327/job/37745896045?pr=7622#step:12:27) on dune app, ensure you are happy with results?

bowenli86 commented:
@jeff-dude tried, fixed a casting issue, made a couple into tables as they seem very expensive to query repeatedly, but feel free to revert if you disagree

jeff-dude commented:
> @jeff-dude tried, fixed a casting issue, made a couple into tables as they seem very expensive to query repeatedly, but feel free to revert if you disagree

those changes work for me. if you're good, then i'm good. keep in mind, if these tables really grow, it would be good to consider making them incremental instead. right now, they are small enough to build quickly on full history each time. let's keep this in mind as you continue to build in future PRs.

bowenli86 commented:
> > @jeff-dude tried, fixed a casting issue, made a couple into tables as they seem very expensive to query repeatedly, but feel free to revert if you disagree
> 
> those changes work for me. if you're good, then i'm good. keep in mind, if these tables really grow, it would be good to consider making them incremental instead. right now, they are small enough to build quickly on full history each time. let's keep this in mind as you continue to build in future PRs.

sounds good, thank you!

bowenli86 commented:
@jeff-dude lmk if this can be merged today
